### PR TITLE
Add way to skip only actual APIs

### PR DESCRIPTION
**Pull request by romanowski**
Now instead of skipping whole APIs we skip only actual APIs.

~~I will be AFK for a while (up to ScalaWorld) so I fear I won't be able to update this PR (and we need to add at least one test for it).~~

### CODE REVIEW

**Comment by typesafe-tools** [comment]
To validate this pull request against other sbt modules, please visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt%20validator/parambuild/?zinc=pull/399/head&deploy_to_bintray=false">this link.</A>

**Comment by typesafe-tools** [comment]
To validate this pull request against other sbt modules, please visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild/?zinc=pull/399/head&deploy_to_bintray=false&run_sbt_community=false">this link.</A>

**Comment by typesafe-tools** [comment]
To validate this pull request against other sbt modules, please visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild/?zinc=pull/399/head&deploy_to_bintray=false&run_sbt_community=false">this link.</A>

**Comment by romanowski** [comment]
@jvican @eed3si9n My changes are not binary compatible however I don't see a clean and non-hacky way to fix the bug that will preserve it.

We need to either fix the bug or remove 'storeApis' flag (but now we don't need apis for core zinc functionality so I think we should keep supporting it).

**Comment by eed3si9n** [comment]
@romanowski Binary compatibility of what exactly? The internal class signatures or the protobuf files? If it's for internal classes like the ones in this PR, we should be the only people using them, so I think we are free to break them. Just add mima exclusion settings.

**Comment by lightbend-cla-validator** [comment]
Hi @invalid-email-address,

Thank you for your contribution! We really value the time you've taken to put this together.

Before we proceed with reviewing this pull request, please sign the Lightbend Contributors License Agreement:

[http://www.lightbend.com/contribute/cla](http://www.lightbend.com/contribute/cla)

**Comment by typesafe-tools** [comment]
To validate this pull request against other sbt modules, please visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild/?zinc=pull/399/head&deploy_to_bintray=false&run_sbt_community=false">this link.</A>

**Comment by typesafe-tools** [comment]
To validate this pull request against other sbt modules, please visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild/?zinc=pull/399/head&deploy_to_bintray=false&run_sbt_community=false">this link.</A>

File: build.sbt [hunk 1]
```diff
@@ -198,6 +198,21 @@ lazy val zincCompile = (project in file("zinc-compile"))
   )
   .configure(addSbtUtilTracking)
 
+def zincPersistMimaExcludes = {
```
Comment (romanowski) [inline build.sbt]: I am not sure if this is the best place to keep excludes so I am open for suggestions.

Comment (jvican) [reply build.sbt]: Can you make it a mima exclude file as they have in the compiler for Scala reflect and the library?

**Comment by jvican** [comment]
Hey @romanowski, thanks for the follow-up. Tomorrow I'll review this. Can you add tests to this?

**Comment by typesafe-tools** [comment]
A validation involving this pull request <a href="https://jenkins.scala-sbt.org:8592/job/sbt-validator/189/console">is in progress...</a>

**Comment by typesafe-tools** [comment]
[//]: # (Refs: 1.x-pull/399/head-1.x-1.x-1.x-1.x)
The validator <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/189/">has checked</A> the following projects against Scala 2.12,
tested using dbuild, projects built on top of each other.

Project             | Reference            | Commit
:------------------ | :------------------- | :----------------
  sbt               | <A HREF="https://github.com/sbt/sbt/tree/1.x">1.x                 </A> | sbt/sbt@4b8d31a9ea861e14922c640d1aa3935a7f33a8fb
  zinc              | <A HREF="https://github.com/sbt/zinc/pull/399">pull/399/head       </A> | sbt/zinc@15f4f99fc4c6dbcdeab8276ae1dd3b55866356f1
  io                | <A HREF="https://github.com/sbt/io/tree/1.x">1.x                 </A> | sbt/io@62004b28dfcdc33273ebb07a563e78178f6eb21d
  librarymanagement | <A HREF="https://github.com/sbt/librarymanagement/tree/1.x">1.x                 </A> | sbt/librarymanagement@293666f653d2686e3049dce37fb80ccb58845b99
  util              | <A HREF="https://github.com/sbt/util/tree/1.x">1.x                 </A> | sbt/util@cba6f83447b84cc1e967f25a930d4f411f6f0664
  website           | <A HREF="https://github.com/sbt/website/tree/1.x">1.x                 </A> | 

:white_check_mark: The result is: ***SUCCESS***
<A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild?sbt=1.x&zinc=pull/399/head&io=1.x&librarymanagement=1.x&util=1.x&website=1.x&deploy_to_bintray=false&run_sbt_community=false">(restart)</A>

**Comment by romanowski** [comment]
Sorry for long delay but you know how life is. Tests added (and it requires some new features in `IncHandler`).

**Comment by jvican** [comment]
Don't worry @romanowski. I've updated this branch manually, if tomorrow CI is green, I'll LGTM. Then I'll leave to the sbt team to approve and merge this change.

File: internal/zinc-persist/src/main/scala/sbt/internal/inc/binary/converters/ProtobufReaders.scala [hunk 2]
```diff
@@ -531,7 +531,15 @@ final class ProtobufReaders(mapper: ReadMapper) {
     import ReadersFeedback.ExpectedCompanionsInAnalyzedClass
     val compilationTimestamp = analyzedClass.compilationTimestamp
     val name = analyzedClass.name
-    val api = mkLazy(analyzedClass.api.read(fromCompanions, ExpectedCompanionsInAnalyzedClass))
+    val api =
+      if (storeApis)
+        mkLazy((analyzedClass.api.read(fromCompanions, ExpectedCompanionsInAnalyzedClass)))
+      else
+        new Lazy[Companions] {
```
Comment (jvican) [inline internal/zinc-persist/src/main/scala/sbt/internal/inc/binary/converters/ProtobufReaders.scala]: Can we move this class into a private local object defined outside of here?

File: build.sbt [hunk 3]
```diff
@@ -207,7 +207,7 @@ lazy val zincPersist = (project in internalPath / "zinc-persist")
     libraryDependencies += sbinary,
     compileOrder := sbt.CompileOrder.Mixed,
     PB.targets in Compile := List(scalapb.gen() -> (sourceManaged in Compile).value),
-    mimaSettings,
```
Comment (jvican) [inline build.sbt]: Can you bring it back?

File: internal/zinc-persist/src/main/mima-filters/1.0.0.backwards.excludes [hunk 4]
```diff
@@ -0,0 +1,8 @@
+ProblemFilters.exclude[DirectMissingMethodProblem]("sbt.internal.inc.binary.BinaryAnalysisFormat.readAPIs")
```
Comment (jvican) [inline internal/zinc-persist/src/main/mima-filters/1.0.0.backwards.excludes:1]: Just an appreciation: this means that we **do not** provide bincompat for readers and writers, **only for the protobuf format**.

File: internal/zinc-persist/src/main/scala/sbt/internal/inc/binary/BinaryAnalysisFormat.scala [hunk 5]
```diff
@@ -38,10 +39,12 @@ final class BinaryAnalysisFormat(mappers: ReadWriteMappers) {
     analysis -> miniSetup
   }
 
-  def readAPIs(reader: CodedInputStream, analysis0: CompileAnalysis): CompileAnalysis = {
+  def readAPIs(reader: CodedInputStream,
+               analysis0: CompileAnalysis,
+               storeApis: Boolean): CompileAnalysis = {
```
Comment (jvican) [inline internal/zinc-persist/src/main/scala/sbt/internal/inc/binary/BinaryAnalysisFormat.scala]: Nitpick: can we rename all instances of `storeApis` into `shouldStoreApis`?

File: internal/zinc-persist/src/main/scala/sbt/internal/inc/binary/converters/ProtobufReaders.scala [hunk 6]
```diff
@@ -531,7 +531,15 @@ final class ProtobufReaders(mapper: ReadMapper) {
     import ReadersFeedback.ExpectedCompanionsInAnalyzedClass
     val compilationTimestamp = analyzedClass.compilationTimestamp
     val name = analyzedClass.name
-    val api = mkLazy(analyzedClass.api.read(fromCompanions, ExpectedCompanionsInAnalyzedClass))
+    val api =
+      if (storeApis)
```
Comment (jvican) [inline internal/zinc-persist/src/main/scala/sbt/internal/inc/binary/converters/ProtobufReaders.scala]: Nitpick: can you put the `if` into two lines?

File: internal/zinc-persist/src/main/scala/sbt/internal/inc/binary/converters/ProtobufWriters.scala [hunk 7]
```diff
@@ -655,14 +655,14 @@ final class ProtobufWriters(mapper: WriteMapper) {
     )
   }
 
-  def toApis(apis: APIs): schema.APIs = {
-    val internal = apis.internal.mapValues(toAnalyzedClass)
-    val external = apis.external.mapValues(toAnalyzedClass)
+  def toApis(apis: APIs, storeApis: Boolean): schema.APIs = {
+    val internal = apis.internal.mapValues(toAnalyzedClass(storeApis))
```
Comment (jvican) [inline internal/zinc-persist/src/main/scala/sbt/internal/inc/binary/converters/ProtobufWriters.scala]: `toAnalyzedClass(storeApis)` should be hoisted up in a local val.

File: internal/zinc-scripted/src/test/scala/sbt/internal/inc/IncHandler.scala [hunk 8]
```diff
@@ -349,26 +354,32 @@ case class ProjectStructure(
...
-                               optionProgress = None,
-                               extra)
+    val previousResult = prev(
+      Option(properties.getProperty("alwaysLoadAnalysis")).exists(_.toBoolean))
+    val setup = compiler.setup(
+      lookup,
+      skip = false,
+      cacheFile,
+      cache = CompilerCache.fresh,
+      incOptions.withClassfileManagerType(transactional).withStoreApis(storeApis),
```
Comment (jvican) [inline internal/zinc-scripted/src/test/scala/sbt/internal/inc/IncHandler.scala]: Since you're in this area, can you move this to a local val?

**Review by jvican — CHANGES_REQUESTED** [review]
Hey @romanowski, thanks for rebasing. I have some questions that I drop you here.
  
Also, this PR adds a new user-facing option called `cachedAnalysis`. Why was that added? Motivation?

**Comment by romanowski** [comment]
> Also, this PR adds a new user-facing option called cachedAnalysis. Why was that added? Motivation?

I added `useCachedAnalysis` flg to IncHandler that is used only in our scripted tests. You need to turn to false if you want test binary format since but default IncHandler use analysis cached in memory. I've also pushed required changes.

FYI this is critial buf for us and currently we are using custom version with this fix on production.

**Comment by dwijnand** [comment]
@jvican Are you happy with the pushed changes and answer? How do you feel about this PR now?

**Review by jvican — DISMISSED** [review]
This is good to go! Thank you @romanowski.

**Review by jvican — APPROVED** [review]

**Comment by romanowski** [comment]
Thanks!

**Pull request merged**